cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java

```diff
@@ -100,8 +106,27 @@ public void set(BasicMapReduceContext context, CConfiguration conf, Transaction
     setPlugins(context.getApplicationSpecification().getPlugins());
     setProgramJarURI(programJarURI);
     setConf(conf);
-    setTx(tx);
     setLocalizedResources(localizedUserResources);
+    setOutputs(context.getOutputs());
+  }
+
+  private void setOutputs(List<ProvidedOutput> providedOutputs) {
+    // we only need to serialize the original Output objects, not the entire ProvidedOutput
+    List<Output.DatasetOutput> datasetOutputs = new ArrayList<>();
+    for (ProvidedOutput providedOutput : providedOutputs) {
+      Output output = providedOutput.getOutput();
+      if (output instanceof Output.DatasetOutput) {
```

Comment:
Right. If someone instantiates the dataset and then calls `addOutput(Output.of(String outputName, OutputFormatProvider outputFormatProvider))`, we'll disallow that.
They should use the `Output.ofDataset(String datasetName, Map<String, String> args)`, since there's no guaranteed way for us to serialize `OutputFormatProvider` objects.